core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java

```diff
@@ -158,39 +162,138 @@ private B2FileVersion finishLargeFileFromB2PartFutures(B2FileVersion largeFileVe
+                                new B2ContentSourceWithByteProgressListener(contentSource, progressListener);
+                        final B2UploadPartRequest uploadPartRequest = B2UploadPartRequest
+                                .builder(partNumber, contentSourceThatReportsProgress)
+                                .build();
+
+                        updateProgress(
+                                uploadListener,
+                                partNumber,
+                                contentSource.getContentLength(),
+                                0,
+                                B2UploadState.STARTING);
+
+                        final B2Part part = webifier.uploadPart(uploadPartUrlResponse, uploadPartRequest);
+
+                        // Return the upload part URL, because it works and can be reused.
+                        uploadPartUrlCache.unget(uploadPartUrlResponse);
+
+                        updateProgress(
+                                uploadListener,
+                                partNumber,
+                                part.getContentLength(),
+                                part.getContentLength(),
+                                B2UploadState.SUCCEEDED);
+
+                        return part;
+                    },
+                    retryPolicySupplier.get()
+            );
+        } catch (Exception e) {
+            updateProgress(
+                    uploadListener,
+                    partNumber,
+                    contentSource.getContentLength(),
+                    0,
+                    B2UploadState.FAILED);
+
+            throw e;
+        }
+    }
 
-                    // Return the upload part URL, because it works and can be reused.
-                    uploadPartUrlCache.unget(uploadPartUrlResponse);
+    B2Part copyPart(
+            int partNumber,
+            String sourceFileId,
+            B2ByteRange byteRangeOrNull,
+            B2UploadListener uploadListener) throws B2Exception {
 
-                    return part;
-                },
-                retryPolicySupplier.get()
-        );
-    }
+        updateProgress(
+                uploadListener,
+                partNumber,
+                1,
+                0,
+                B2UploadState.WAITING_TO_START);
 
-    B2Part copyPart(int partNumber, String sourceFileId, B2ByteRange byteRangeOrNull) throws B2Exception {
         final B2CopyPartRequest copyPartRequest = B2CopyPartRequest
                 .builder(partNumber, sourceFileId, fileVersion.getFileId())
                 .setRange(byteRangeOrNull)
                 .build();
 
-        return retryer.doRetry(
-                "b2_copy_part",
-                accountAuthCache, () -> webifier.copyPart(accountAuthCache.get(), copyPartRequest),
-                retryPolicySupplier.get());
+        try {
+            return retryer.doRetry(
+                    "b2_copy_part",
+                    accountAuthCache, () -> {
```

Comment:
nitpick?  don't start the lambda on the same line as the previous argument?

Reply:
Done.